precompiles/utils/src/precompile_set.rs

```diff
@@ -18,241 +18,389 @@
 //! final precompile set with security checks. All security checks are enabled by
 //! default and must be disabled explicely throught type annotations.
 
-use crate::{revert, StatefulPrecompile};
+use crate::{revert, substrate::RuntimeHelper};
 use fp_evm::{Precompile, PrecompileHandle, PrecompileResult, PrecompileSet};
 use frame_support::pallet_prelude::Get;
 use impl_trait_for_tuples::impl_for_tuples;
 use pallet_evm::AddressMapping;
-use sp_core::H160;
+use sp_core::{H160, H256};
 use sp_std::{
 	cell::RefCell, collections::btree_map::BTreeMap, marker::PhantomData, ops::RangeInclusive, vec,
 	vec::Vec,
 };
 
-// CONFIGURATION TYPES
-
 mod sealed {
 	pub trait Sealed {}
 }
 
-/// How much recursion is allows for a precompile.
-pub trait RecursionLimit: sealed::Sealed {
-	fn recursion_limit() -> Option<u16>;
-}
+/// Trait representing checks that can be made on a precompile call.
+/// Types implementing this trait are made to be chained in a tuple.
+/// For that reason every method returns an Option, None meaning that
+/// the implementor have no constraint and the decision is left to
+/// latter elements in the chain. If None is returned by all elements of
+/// the chain then sensible defaults are used.
+pub trait PrecompileChecks {
```

Comment:
This is a bit unclear (or just complicated), maybe an example would help. Alternatively, if it's not meant to be used by e.g. individual precompiles, it might help to explain where it is/isn't used

Reply:
Adding a bit more info on where it is used.